File: apps/android/app/src/main/java/com/sennnen/mav/ui/mav/MavNarrative.kt
package com.sennnen.mav.ui.mav

import com.sennnen.mav.BuildConfig

// The written half of Today. The iOS twin is Model/MavNarrative.swift.
//
// This is the one surface where placeholder copy is allowed, and it is fenced: sample text exists
// only in a debug build and is always badged on screen. A release build with no model wired shows
// the honest not-yet-generated state. On-device generation (Gemini Nano here, Foundation Models on
// iOS) and bring-your-own-key advisor chat are a later lane.

sealed interface MavNarrativeState {
    /** Written on-device from the day's own read models. */
    data class Generated(val title: String, val text: String) : MavNarrativeState

    /** Fixture copy. Debug builds only, always rendered behind a visible SAMPLE badge. */
    data class Sample(val title: String, val text: String) : MavNarrativeState

    /** Nothing has been generated. Says so. */
    data class Unavailable(val reason: String) : MavNarrativeState

    val headline: String?
        get() = when (this) {
            is Generated -> title
            is Sample -> title
            is Unavailable -> null
        }

    val bodyText: String?
        get() = when (this) {
            is Generated -> text
            is Sample -> text
            is Unavailable -> null
        }

    val isSample: Boolean get() = this is Sample
}

/** Keeping this an interface is what lets the model lane land later without touching a screen. */
interface MavNarrativeProvider {
    fun daily(day: String, rows: List<MavMetricRow>): MavNarrativeState
    fun trend(id: String, rows: List<MavMetricRow>): MavNarrativeState
}

object MavStubNarrativeProvider : MavNarrativeProvider {
    const val NOT_GENERATED_YET =
        "On-device summaries are not wired up yet. When they are, this is where the day gets " +
            "explained in words, written on your phone from your own data."

    override fun daily(day: String, rows: List<MavMetricRow>): MavNarrativeState =
        if (BuildConfig.MAV_SHOW_SYNTHETIC_DATA) {
            MavNarrativeState.Sample(
                title = "Your resting pulse is trending lower",
                text = "Your three-week average is down while overnight variability remains steady.",
            )
        } else {
            MavNarrativeState.Unavailable(NOT_GENERATED_YET)
        }

    override fun trend(id: String, rows: List<MavMetricRow>): MavNarrativeState = when {
        !BuildConfig.MAV_SHOW_SYNTHETIC_DATA -> MavNarrativeState.Unavailable(NOT_GENERATED_YET)
        id == "resilience" -> MavNarrativeState.Sample(
            "Recovery has stayed consistent over the past month.", "",
        )
        id == "cardio_load" -> MavNarrativeState.Sample(
            "Training load has increased gradually over eight weeks.", "",
        )
        else -> MavNarrativeState.Unavailable(NOT_GENERATED_YET)
    }
}

File: apps/ios/Maverick/Model/MavNarrative.swift
import Foundation

// The written half of `Today`.
//
// This is the one surface in the app where placeholder copy is allowed, and it is fenced: sample
// text exists only in a debug build, and it is always badged on screen. A release build with no
// model wired shows the honest not-yet-generated state. On-device generation (Foundation Models on
// iOS, Gemini Nano on Android) and bring-your-own-key advisor chat are a later lane; this is the
// surface and the empty state, nothing more.

enum MavNarrativeState: Equatable, Sendable {
  /// Written on-device from the day's own read models.
  case generated(headline: String, body: String)
  /// Fixture copy. Debug builds only, and always rendered behind a visible SAMPLE badge.
  case sample(headline: String, body: String)
  /// Nothing has been generated. Says so.
  case unavailable(reason: String)

  var headline: String? {
    switch self {
    case .generated(let headline, _), .sample(let headline, _): headline
    case .unavailable: nil
    }
  }

  var body: String? {
    switch self {
    case .generated(_, let body), .sample(_, let body): body
    case .unavailable: nil
    }
  }

  var isSample: Bool {
    if case .sample = self { return true }
    return false
  }
}

/// What a narrative is asked for. Keeping this a protocol is what lets the model lane land later
/// without touching a screen.
protocol MavNarrativeProviding: Sendable {
  func daily(day: String, rows: [MavMetricRow]) -> MavNarrativeState
  func trend(id: String, rows: [MavMetricRow]) -> MavNarrativeState
}

/// The provider until a model is wired.
///
/// In release it returns `.unavailable` for everything, which is the truth. In debug it returns
/// fixture copy so the layout can be judged — and the screen badges it, so a screenshot can never
/// be mistaken for a working feature.
struct MavStubNarrativeProvider: MavNarrativeProviding {
  static let notGeneratedYet =
    "On-device summaries are not wired up yet. When they are, this is where the day gets explained "
    + "in words, written on your phone from your own data."

  func daily(day: String, rows: [MavMetricRow]) -> MavNarrativeState {
    #if DEBUG
      return .sample(
        headline: "Your resting pulse is trending lower",
        body: "Your three-week average is down while overnight variability remains steady.")
    #else
      return .unavailable(reason: Self.notGeneratedYet)
    #endif
  }

  func trend(id: String, rows: [MavMetricRow]) -> MavNarrativeState {
    #if DEBUG
      switch id {
      case "resilience":
        return .sample(
          headline: "Recovery has stayed consistent over the past month.", body: "")
      case "cardio_load":
        return .sample(
          headline: "Training load has increased gradually over eight weeks.", body: "")
      default:
        return .unavailable(reason: Self.notGeneratedYet)
      }
    #else
      return .unavailable(reason: Self.notGeneratedYet)
    #endif
  }
}
